src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/security/providers/okhttpMontaguAPIClient.kt

```diff
@@ -0,0 +1,129 @@
+import com.github.salomonbrys.kotson.fromJson
+import com.google.gson.JsonSyntaxException
+import okhttp3.Headers
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import org.slf4j.LoggerFactory
+import org.vaccineimpact.orderlyweb.Serializer
+import org.vaccineimpact.orderlyweb.db.AppConfig
+import org.vaccineimpact.orderlyweb.db.Config
+import java.io.IOException
+import java.security.cert.X509Certificate
+import javax.net.ssl.*
+
+
+interface MontaguAPIClient
+{
+    @Throws(MontaguAPIException::class)
+    fun getUserDetails(token: String): UserDetails
+
+    data class UserDetails(val email: String, val username: String, val name: String?)
+
+    // The following are identical to orderlyweb.models.Result, orderlyweb.models.ResultStatus,
+    // and orderlyweb.models.ErrorInfo but in principle they need not be and should either spec
+    // change could diverge, so defining Montagu specific models here
+    data class Result(val status: String, val data: Any?, val errors: List<ErrorInfo>)
+
+    data class ErrorInfo(val code: String, val message: String)
+    {
+        override fun toString(): String = message
+    }
+
+}
+
+class okhttpMontaguAPIClient(appConfig: Config = AppConfig()) : MontaguAPIClient
+{
+    private val urlBase = appConfig["montagu.api_url"]
+    private val devMode = appConfig["proxy.dev.mode"]
+    private val serializer = Serializer.instance.gson
+
+    private val logger = LoggerFactory.getLogger(okhttpMontaguAPIClient::class.java)
+
+    override fun getUserDetails(token: String): MontaguAPIClient.UserDetails
+    {
+        getHttpResponse("$urlBase/user/", mapOf("Authorization" to "Bearer $token"))
+                .use { response ->
+                    val result = parseResult(response.body!!.string())
+
+                    if (response.code != 200)
+                    {
+                        throw MontaguAPIException("Response had errors ${result.errors.joinToString(",") { it.toString() }}", response.code)
+                    }
+
+                    return result.data as MontaguAPIClient.UserDetails
+                }
+    }
+
+    private fun parseResult(jsonAsString: String): MontaguAPIClient.Result
+    {
+        return try
+        {
+            val result = serializer.fromJson<MontaguAPIClient.Result>(jsonAsString)
+            if (result.data.toString().isNotEmpty())
+            {
+                result.copy(data = serializer.fromJson<MontaguAPIClient.UserDetails>(serializer.toJson(result.data)))
+            }
+            else result
+        }
+        catch (e: JsonSyntaxException)
+        {
+            throw MontaguAPIException("Failed to parse text as JSON.\nText was: $jsonAsString\n\n$e", 500)
+        }
+
+    }
+
+
+    private fun getHttpResponse(url: String, headers: Map<String, String>): Response
+    {
+        val client: OkHttpClient;
+        if (devMode != "true")
+        {
+            client = OkHttpClient()
+        }
+        else
+        {
```

Comment:
I think it would be tidier to move this logic out of this class, rather have a method `protected open getClient` and then create an inherited class `DevModeMontaguAPIClient` that just overrides `getClient` with the logic inside this else block. Then choose which client we use in `MontaguIndirectClient` based on the config value?

Or at the very least move this logic out into 2 methods, `getClient` and `getLocalhostClient`/`getDevModeClient()` or something like that, to keep this method easier to follow?